Support for named strings in margin-boxes

E-M-P-I-R-E opened the pull request:
Named strings using the `-weasy-string-set` style and the CSS `string` function passed to `content()`.
http://dev.w3.org/csswg/css-gcpm/#named-strings
![named_strings](https://cloud.githubusercontent.com/assets/2049558/6745711/b3f9caaa-ce91-11e4-8974-60ed1d4f2d5e.png)

More discussion in feature request - issue https://github.com/Kozea/WeasyPrint/issues/174

Below is a simple example (more in tests/test_boxes.py).

``` html
<html>
<head>
<style>
body{
     counter-reset: header_counter; 
}
@page:first{
    @top-center { content: string(header)}
}
@page{
    @top-center { content: string(header, first-except)}
}
.page{
      page-break-before: always;
}

h1{
    counter-increment: header_counter;
	string-set: header content(after);
	-weasy-string-set: header content(text);
}
h1:before{
	content: "Chapter " counter(header_counter) " ";
}
</style>
</head>
<body>
    <div class="page"><h1>First</h1>This demonstrates named strings using the `string-set` style and the CSS `string` function passed to `content()`.<br>There are two assignments "First" and "Second" plus the psuedo ::before element. The numbers are produced using CSS `counter`.</div>
    <div class="page">CONTENT</div>
    <div class="page" ><h1 >Second</h1>Pages after the first have `first-except` set, meaning the page on which the assignment is made is excluded. Proceeding pages will have the value set here ("Second").</div>
    <div class="page">CONTENT</div>
</body>
</html>
```

SimonSapin commented:
This looks great! I’ll comment on specifics inline.

Tests [fail](https://travis-ci.org/Kozea/WeasyPrint/jobs/55126940) on Travis-CI because:

* `flake8` complains about whitespace and overlong lines
* `xrange` doesn’t exist on Python 3.x. Add `from ..compat import xrange` at the top of the file.

SimonSapin commented on weasyprint/css/properties.py:
```diff
@@ -97,6 +97,8 @@
     'position': 'static',
     'right': 'auto',
+    #default not defined in spec http://dev.w3.org/csswg/css-gcpm/#named-strings
+    'string_set': '',
```
The initial value *is* defined in the spec, in the property definition table. It is `none`, which should be represented as a string `'none'` since returning `None` in `validation.py` indicates a parse error.

SimonSapin commented on weasyprint/css/validation.py:
```diff
@@ -685,6 +685,8 @@ def validate_content_token(base_url, token):
             if style in ('none', 'decimal') or style in counters.STYLES:
                 return (name, args)
+        elif prototype in (('string', ['IDENT']),('string', ['IDENT', 'IDENT'])):
+            return (name, args)
```
When there are two arguments, this should check that the second one is one of the supported keywords.

SimonSapin replied:
These keywords should also be case-insensitive. (But not the string/variable’s name!)

SimonSapin commented on weasyprint/css/validation.py:
```diff
@@ -1267,6 +1269,23 @@ def bookmark_level(token):
 
 
+@validator(prefixed=True)  # CSS3 GCPM, experimental
+def string_set(tokens):
```
I have a hard time following this, sorry. Could you clarify what set of syntax you mean to support? All unsupported syntax should be rejected.

SimonSapin commented on weasyprint/css/validation.py:
```diff
@@ -1267,6 +1269,23 @@ def bookmark_level(token):
+        raise InvalidValues("Only one value is supported. Received %s" %len(tokens)-1)
+    function = parse_function(tokens[1])
+    if function is None:
+        keyword = 'text'
```
When the second token is not a function, it could be anything else. This should check what it is and only accept the expected syntax.

SimonSapin commented on weasyprint/css/validation.py:
```diff
@@ -1267,6 +1269,23 @@ def bookmark_level(token):
+    if function is None:
+        keyword = 'text'
+    else:
+        name, args = function
```
`name` should be checked here.

SimonSapin commented on weasyprint/css/validation.py:
```diff
@@ -1267,6 +1269,23 @@ def bookmark_level(token):
+        name, args = function
+        keyword = args[0].value
+    var_name = get_keyword(tokens[0])
+    if tokens[0].type == "IDENT" and keyword in ('text', 'before', 'after'):
```
Keywords should be case-insensitive. (Use `.lower()`.)

SimonSapin commented on weasyprint/css/validation.py:
```diff
@@ -1267,6 +1269,23 @@ def bookmark_level(token):
+        keyword = args[0].value
+    var_name = get_keyword(tokens[0])
+    if tokens[0].type == "IDENT" and keyword in ('text', 'before', 'after'):
+        return ('keyword', keyword, 'name', var_name)
```
Do the `'keyword'` and `'name'` strings add information here, or could the return value be just `(keyword, name)`? (Or `(name, keyword)`, to match the order of the syntax?)

SimonSapin commented on weasyprint/formatting_structure/build.py:
```diff
@@ -235,6 +236,9 @@ def content_to_boxes(style, parent_box, quote_depth, counter_values,
                 for counter_value in counter_values.get(counter_name, [0])
             ))
+        elif type_ == 'string':
+            text = context.get_string_set_for(*tuple(value))
```
Avoid crashing here when `context` is `None`. (Maybe just ignore `string()` functions.)

SimonSapin replied:
`*tuple(value)` can be just `*value`

SimonSapin commented on weasyprint/layout/__init__.py:
```diff
@@ -71,6 +70,8 @@ def __init__(self, enable_hinting, style_for, get_image_from_uri):
         self.get_image_from_uri = get_image_from_uri
         self._excluded_shapes_lists = []
         self.excluded_shapes = None  # Not initialized yet
+        self.string_set = defaultdict(lambda:defaultdict(lambda:[]))
```
Nit: replace `lambda: []` with `list`.

SimonSapin commented on weasyprint/layout/__init__.py:
```diff
@@ -89,3 +90,33 @@ def finish_block_formatting_context(self, root_box):
+            Value depends on current page - http://dev.w3.org/csswg/css-gcpm/#funcdef-string
+
+            :param name: the name of the named string.
+            :param name: indicates which value of the named string should be used.
```
Should this be `:param keyword:`?

SimonSapin commented on weasyprint/layout/__init__.py:
```diff
@@ -89,3 +90,33 @@ def finish_block_formatting_context(self, root_box):
+
+        """
+        last = keyword == "last"
+        if self.current_page in self.string_set[name]:
```
Does this crash if `name` is not in `self.string_set`? Maybe replace all occurrences of `self.string_set[name]` with a single `assignments = self.string_set.get(name)` and check for `assignments is None`.

E-M-P-I-R-E replied:
It doesn't crash. The way the defaultdict is structured and evaluated [when resolving the value] is checking the current page for assignments, then previous pages... finally returning an empty string if there were no assignments.

SimonSapin replied:
Oh, right, I forgot it was a defaultdict.

SimonSapin commented on weasyprint/layout/__init__.py:
```diff
@@ -89,3 +90,33 @@ def finish_block_formatting_context(self, root_box):
+            for previous_page in xrange(self.current_page, 0, -1):
+                if previous_page in self.string_set[name]:
+                    return self.string_set[name][previous_page][-1]
+        return ""
```
Yeah, the spec doesn’t say what to do in this case, but I suppose this is sensible.

E-M-P-I-R-E replied:
I'm not sure exactly which case you're referring to.

SimonSapin commented on weasyprint/layout/pages.py:
```diff
@@ -453,8 +454,19 @@ def page_width(box, context, containing_block_width):
+    if isinstance(box, boxes.TextBox):
+        return box.text
+    elif isinstance(box, boxes.ParentBox):
+        return ''.join(box_text_contents(child) for child in box.children)
```
Although that is not the common case, this code will create lots of intermediate strings if the tree is deep. Please this line with:

```
return ''.join(b.text for b in box.descendants() if isinstance(b, boxes.TextBox))
```
(

SimonSapin commented on weasyprint/layout/pages.py:
```diff
@@ -511,6 +523,13 @@ def make_page(context, root_box, page_type, resume_at, content_empty):
     context.finish_block_formatting_context(root_box)
 
     page = page.copy_with_children([root_box])
+    descendants = list(page.descendants())
```
This list might consume a lot of memory and doesn’t seem to serve any purpose. Please use the generator directly in the for-loop: `for child in page.descendants()`:.

SimonSapin commented:
Alright, I’m done reviewing up to 7db9be4005b2188b9674d4baa0b832afd67d88dc. There are some minor issues, but the overall design is simpler than I expected. Good job!

E-M-P-I-R-E commented:
Oops, I guess I should have responded to some of your comments before pushing again [I was planning on it]. I think I addressed everything in this latest commit. Mostly, changed the validation.

One oddity - you suggested a change  in `TEXT_CONTENT_EXTRACTORS['box_text_contents']`. I wasn't actually using that function (meant to remove that code), so I made that change to the original function in `formatting_structure/build`. I also added a new key `text` which calls that function.

Oh.. I just figured out how to see the specific issues that are causing continuous integration to fail. I'll take care of that.

Two things that I'm unclear about
First - from the spec.:

> User agents must be able to recall many values of the named string, as the string() function can return past, current, or **future values** of the assignment.

Second: should there be a warning when attempting to use named strings outside of the margin-boxes?

Thanks a lot!

E-M-P-I-R-E commented:
Question was raised:
>Do the 'keyword' and 'name' strings add information here, or could the return value be just (keyword, name)? (Or (name, keyword), to match the order of the syntax?)

I was following the example of the validation function for `bookmark-label` [which returns `('string', token.value)`] but I liked your suggestion and went with that.

`string-set` validation now returns `(var_name, keyword)`. It's only accessed in `layout/pages.make_page()` (when updating the `LayoutContext` with new assignments for a given named-string).

E-M-P-I-R-E commented:
It was also requested that I explain what I intended to support for `string-set` (in terms of validation).

The answer is one value pair ex:
`-weasy-string-set: header content();`

If no argument is passed to `content()` it defaults to 'text'.

liZe commented:
Wow. So cool!

I'll review that and merge it soon. Thanks @E-M-P-I-R-E for your work, and @SimonSapin for your review :).
